src/task/cluster.rs

```diff
@@ -497,6 +497,76 @@ impl Cluster {
     pub fn failover(&self) -> &Failover {
         &self.failover
     }
+
+    pub fn try_upgrade(mut self, new: &Cluster) -> Result<Self, GeninError> {
+        let old_hosts = self.hosts().first().unwrap().lower_level_hosts();
+        let new_hosts = new.hosts().first().unwrap().lower_level_hosts();
+
+        let mut diff = new_hosts
+            .into_iter()
+            .flat_map(|new_host| {
+                new_host.instances.iter().filter(|new_instance| {
+                    !old_hosts.iter().any(|old_host| {
+                        old_host
+                            .instances
+                            .iter()
+                            .any(|old_instance| old_instance.name.eq(&new_instance.name))
+                    })
+                })
+            })
+            .collect::<Vec<&InstanceV2>>();
+
+        diff.sort();
+
+        println!(
+            "New instances: {}",
+            diff.iter()
+                .map(|instance| instance.name.to_string())
+                .collect::<Vec<String>>()
+                .join(" ")
+        );
+
+        self.failover = new.failover.clone();
+        self.vars = new.vars.clone();
+
+        if let Some(host) = self.hosts.first_mut() {
+            host.delete_stateboard();
```

Comment:
can you explain, why is stateboard always placed to the end?

and I don't understand, what will happen if old and new cluster have etcd2 failover?

Reply:
I did this because the stateboard is the same instance as all the others, and whatever it takes up the port, and does not cause a shift in distribution, it is always placed at the end. In case etcd2 failover, nothing will happen.